This patch fixes the detection of hidden SSIDs as transmitted
by some cisco systems.

Signed-off-by: Felix Fietkau <nbd@openwrt.org>

--- a/net80211/ieee80211_scan_sta.c
+++ b/net80211/ieee80211_scan_sta.c
@@ -209,6 +209,19 @@
 		ieee80211_saveie(iep, ie);
 }
 
+
+static inline int is_empty_ssid(u_int8_t *ssid)
+{
+	if (!ssid)
+		return 1;
+	if (ssid[1] == 0)
+		return 1;
+	if ((ssid[1] == 1) && (ssid[2] == 0))
+		return 1;
+	return 0;
+}
+
+
 /*
  * Process a beacon or probe response frame; create an
  * entry in the scan cache or update any previous entry.
@@ -252,8 +265,8 @@
 	ise = &se->base;
 
 	/* XXX ap beaconing multiple ssid w/ same bssid */
-	if (sp->ssid[1] != 0 &&
-	    (ISPROBE(subtype) || ise->se_ssid[1] == 0))
+	if (!is_empty_ssid(sp->ssid) &&
+	    (ISPROBE(subtype) || is_empty_ssid(ise->se_ssid)))
 		memcpy(ise->se_ssid, sp->ssid, 2 + sp->ssid[1]);
 
 	memcpy(ise->se_rates, sp->rates, 
